sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -239,15 +240,26 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
     }
   }
 
+  /**
+  * Here use the ReentrantLock instead of synchronized to prevent the dead lock
+  * cause by lazy operator.
+  */
+  private val lock = new ReentrantLock()
+
   /**
    * Blocks the thread until all subqueries finish evaluation and update the results.
    */
-  protected def waitForSubqueries(): Unit = synchronized {
-    // fill in the result of subqueries
-    runningSubqueries.foreach { sub =>
-      sub.updateResult()
+  protected def waitForSubqueries(): Unit = {
```

Comment:
Is it not sufficient to just mark this `synchronized`? is the idea that it would interfere with other synchronization on this instance object?